Feature/HM-97-Properties-Method-that-accepts-model-like-PreferencesModel-and-returns-top-6-props-ids

petarbelchev opened the pull request:
The Properties microservice uses Preferences Store and Recommended Repository.

petarbelchev committed:
The Properties microservice uses Preferences Store and Recommended Repository.

rosteslav commented on src/Properties/Properties.Api/Controllers/PropertiesController.cs:
```diff
@@ -191,7 +191,8 @@ public async Task<IActionResult> Recommended()
 
-            var properties = await _mediator.Send(new GetRecommendedQuery());
+            var userId = User.Claims.First(x => x.Type == ClaimTypes.Sid).Value;
+            var properties = await _mediator.Send(new GetRecommendedQuery() { BuyerId = userId });
```
`(` `)` are not used when with `{` `}` 
In other words 
`new GetRecommendedQuery {BuyerId = userId}` without `()`

rosteslav commented on src/Properties/Properties.Application/Models/PreferencesModel.cs:
```diff
@@ -0,0 +1,20 @@
+namespace BuildingMarket.Properties.Application.Models.Security
+{
+    [MessagePackObject]
+    public class PreferencesModel
```
this exists elsewhere.. if in other microservice, move it to common

rosteslav commented on src/Properties/Properties.Infrastructure/InfrastructureServiceRegistration.cs:
```diff
@@ -22,7 +22,9 @@ public static IServiceCollection AddInfrastructureServices(this IServiceCollecti
 
             services.AddScoped<IPropertiesRepository, PropertiesRepository>();
             services.AddScoped<IPropertyOptionsRepository, PropertyOptionsRepository>();
+            services.AddScoped<IRecommendationRepository, RecommendationRepository>();
```
why not singleton, as well as the other repositories

petarbelchev replied:
I think exposing a repository as a singleton in a concurrent environment can be risky. The DbContext is not thread-safe. It is bad practice to have a single instance of the DbContext for the whole lifetime of the application. It is very important to dispose the DbContext and release the used resources after use?

rosteslav replied:
Yes, that's correct, I didn't consider we have a context in there

rosteslav commented on src/Properties/Properties.Application/Contracts/IPropertiesRepository.cs:
```diff
@@ -10,6 +10,8 @@ public interface IPropertiesRepository
 
         Task<PropertyModel> GetById(int id, CancellationToken cancellationToken);
 
+        Task<IEnumerable<GetAllPropertiesOutputModel>> GetById(IEnumerable<int> ids, CancellationToken cancellationToken);
```
GetByIds, you have more than one

rosteslav commented on src/Properties/Properties.Application/Features/Properties/Queries/GetRecommended/GetRecommendedQueryHandler.cs:
```diff
@@ -4,15 +4,24 @@
+            var properties = await _propertiesRepository.GetById(recommendedIds, cancellationToken);
             if (properties.Any())
             {
                 var propertyIds = properties.Select(p => p.Id.ToString()).ToArray();
```
You don't need that no more. you have it as `recommendedIds` above

petarbelchev replied:
Actually I need this operation (convert to string[]), because the store wants `string[]` and I think it is better to get the ids from the actual properties, instead of the `recommendedIds` collection?

rosteslav replied:
yes, totally

rosteslav commented on src/Properties/Properties.Infrastructure/Repositories/PropertiesRepository.cs:
```diff
@@ -89,6 +88,27 @@ public async Task<PropertyModel> GetById(int id, CancellationToken cancellationT
+            _logger.LogInformation("DB get properties");
+
+            try
+            {
```
check if `ids` is not null or count > 0, no need to go to the db if null or empty

rosteslav commented on src/Properties/Properties.Infrastructure/Repositories/RecommendationRepository.cs:
```diff
@@ -0,0 +1,44 @@
+            try
+            {
+                var properties = await _context.Properties
+                    .OrderBy(p => p.Id)
```
first do a select:
`.Select(p => p.Id)`
`.OrderBy(i => i)`

petarbelchev committed:
Merge branch 'Develop' into feature/HM-97-Properties-Method-that-accepts-model-like-PreferencesModel-and-returns-top-6-props-ids

petarbelchev committed:
Move Preferences Model into Common.
Add Argument Validation to GetByIds Method.
Improve Recommended Properties Query.
